[3.x] Add Integration tests for Resource Prefix

himani2411 opened the pull request:
### Description of changes
Adding Integration Tests Cases for addition of Resource Prefix, where I am creating a different user role with a permission boundary which restricts a user from creating IAM resources without a specific path or name prefix.

-   Add Integration test cases and config file for Iam Resource Prefix in test_iam.py 
-   Add user-role-rp.cfn.yaml to provide a User Role for creating test clusters 
-   Add the test case in pcluster3.yaml config for Jenkins tests.

### Tests
* Manually tested the scenarios covered in integration Tests.

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Commit:
Add Integration tests for Resource Prefix

Add Integration test cases and config file for Iam Resource Prefix in test_iam.py
Add user-role-rp.cfn.yaml to provide a User Role for creating test clusters
Add the test case in pcluster3.yaml config for Jenkins tests.

Commit:
Remove File deleteS3.py

gmarciani commented on tests/integration-tests/tests/iam/test_iam.py:
```diff
@@ -332,3 +336,134 @@ def test_s3_read_write_resource(region, pcluster_config_reader, s3_bucket_factor
+    bucket_name = s3_bucket
+    cfn_client, _, iam_client, _ = _create_boto3_clients(region)
+
+    create_config, update_config = _get_config_create_and_update(test_datadir)
```
The `update_config` variable is never used. 
Is it because the PR is actually a draft and we expect to use it later on?
If this is the case, it would be better to mark the PR as a draft

gmarciani commented on tests/integration-tests/tests/iam/test_iam/test_iam_resource_prefix/pcluster.config.yaml:
```diff
@@ -0,0 +1,49 @@
+  Ssh:
+    KeyName: {{ key_name }}
+  Iam:
+    S3Access:
```
Here and below: why does the cluster need access to a S3 bucket?

himani2411 replied:
I need this as S3 bucket has a few Policies. As you suggested I will test them for Updates to check the updated policies.

gmarciani replied:
Got it, thanks

gmarciani commented on tests/integration-tests/tests/iam/test_iam/test_iam_resource_prefix/pcluster.config.yaml:
```diff
@@ -0,0 +1,49 @@
+            EnableWriteAccess: true
+DevSettings:
+  Timeouts:
+    HeadNodeBootstrapTimeout: 1234  # timeout in seconds
```
Why do we need to set the headnode bootstrap timeout for these tests?

himani2411 replied:
Will remove them.

gmarciani commented on tests/iam_policies/user-role-rp.cfn.yaml:
```diff
@@ -0,0 +1,796 @@
+AWSTemplateFormatVersion: '2010-09-09'
+
```
I suppose `user-role-rp.cfn.yaml` is the copy of [user-role.cfn.yaml](https://github.com/aws/aws-parallelcluster/blob/develop/tests/iam_policies/user-role.cfn.yaml) with the only difference that in `user-role-rp.cfn.yaml` some permissions are restricted to resources (roles and policies) with the expected resource prefix used by the integration tests.

If this is the case I suggest to avoid such duplication because every change to the pcluster user role should then be reflected in two files that must be kept aligned.

What about reading the original yaml file within the test and apply the resource prefix changes at runtime?

himani2411 replied:
I had to make changes by adding a lot of different permissions I the policies as well as the permission boundary as well as removed the Policies related to Image Builder. If I keep these Image Builder policies my File goes over 50Kb and I wont be able to run this template and would have to upload the template through S3 bucket for my Stack to be created.

gmarciani replied:
> If I keep these Image Builder policies my File goes over 50Kb and I wont be able to run this template and would have to upload the template through S3 bucket for my Stack to be created.

Still it's possible to parse the original yaml file and remove the unnecessary resources at runtime to avoid duplication.
Nevertheless, let's proceed iteratively as we agreed on.

the first iteration on integ tests should provide the following values:
1. verify that the feature does not introduce regressions, e.g. having a ResourcePrefix with the standard user role does not break the cluster
1. verify that the ResourcePrefix works in a basic case

himani2411 replied:
Will put this changes in another PR as its non-blocking improvement.

gmarciani commented on tests/integration-tests/tests/iam/test_iam.py:
```diff
@@ -332,3 +336,134 @@ def test_s3_read_write_resource(region, pcluster_config_reader, s3_bucket_factor
+    s3_bucket,
+    iam_resource_prefix,
+):
+    bucket_name = s3_bucket
```
Why do you need a bucket in this test?

himani2411 replied:
I need this as S3 bucket has a Policy. As you suggested I will test them for Updates to check the updated policies.

gmarciani replied:
Got it thanks

gmarciani commented on tests/integration-tests/tests/iam/test_iam.py:
```diff
@@ -332,3 +336,134 @@ def test_s3_read_write_resource(region, pcluster_config_reader, s3_bucket_factor
+            assert_that(resource).contains(iam_name_prefix)
+
+
+def _test_iam_resource_in_cluster(cfn_client, iam_client, stack_name, iam_resource_prefix):
```
Here you're verifying that the IAM resources are created with the expected ResourcePrefix.

I think that these are verifications that could and should be executed at the unit testing level, not at the integration testing level.

What do you think?

himani2411 replied:
Will check for creation of the cluster.

gmarciani replied:
Great and please remember to keep track in a task of the necessity to add a unit tests to cover the generated resources.

dreambeyondorange commented on tests/integration-tests/tests/iam/test_iam.py:
```diff
@@ -332,3 +336,134 @@ def test_s3_read_write_resource(region, pcluster_config_reader, s3_bucket_factor
+    cluster = clusters_factory(cluster_config)
+    _test_iam_resource_in_cluster(cfn_client, iam_client, cluster.name, iam_resource_prefix)
+
+def _split_resource_prefix(resource_prefix):
```
I wonder if we should extract this method from the core implementation so it can be reused here?

himani2411 replied:
I did that originally but during one of the issues with Jenkins which Hanwen helped for solving he suggested I move this in the tests section as the Integration tests should be independent of the core implementation.

dreambeyondorange commented on tests/integration-tests/tests/iam/test_iam.py:
```diff
@@ -332,3 +336,134 @@ def test_s3_read_write_resource(region, pcluster_config_reader, s3_bucket_factor
+
+
+def _test_iam_resource_in_cluster(cfn_client, iam_client, stack_name, iam_resource_prefix):
+    """Test IAM resources ( Roles, policy and Instance profiles)."""
```
Could you elaborate on what exactly we are testing for here?

himani2411 replied:
I am testing if the IAM resources have the expected path and names in their ARN or name.
Apart from that a new change is to add a test for checking the Cluster status, which is what Giacomo suggested [here](https://github.com/aws/aws-parallelcluster/pull/4652#discussion_r1040914842) 

Will add this in the Comments as well.

dreambeyondorange commented on tests/integration-tests/tests/iam/test_iam.py:
```diff
@@ -332,3 +336,134 @@ def test_s3_read_write_resource(region, pcluster_config_reader, s3_bucket_factor
+
+
+@pytest.fixture(scope="class")
+def initialize_resource_prefix_cli_creds(request):
```
This function can be combined with the existing initialize_cli_creds() by adding a parameter or two with default values.  For example
```
stack_template_path = os.path.join("..", "iam_policies", "user-role-rp.cfn.yaml")
```
can be parameterized with `user_role_config_file`

and 
```
stack_name = generate_stack_name(
                    "integ-tests-iam-rp-user-role", request.config.getoption("stackname_suffix")
                )
```
can be parameterized with `base_stack_name`

dreambeyondorange replied:
Passing parameters from a test to a fixture seems overly complicated but can be done using indirect parametrization https://docs.pytest.org/en/7.1.x/example/parametrize.html#indirect-parametrization

himani2411 replied:
The scope of existing initialize_cli_creds() was session and autouse set to True which is why I didnt want to reuse them as the changes I could make have a bigger blast radius and affect the other test cases CLI creds.

gmarciani replied:
I agree that the long term solution would be to make the session-scoped initialize_cli_creds more flexible to be able to also cover this scenario.

Nevertheless, let's proceed iteratively. Let's start simple with a fixture dedicated to our case and improve it in later steps.

dreambeyondorange replied:
good point about the session scope!  I don't really think indirect parametrization will work in that case

dreambeyondorange commented on tests/integration-tests/tests/iam/test_iam.py:
```diff
@@ -332,3 +336,134 @@ def test_s3_read_write_resource(region, pcluster_config_reader, s3_bucket_factor
+
+
+@pytest.fixture(scope="class")
+def register_resource_prefix_cli_credentials(initialize_resource_prefix_cli_creds):
```
this function is identical to the existing `register_cli_credentials`.  Is there a reason you can't just use that function?

himani2411 replied:
The reason is same as [above](https://github.com/aws/aws-parallelcluster/pull/4652#discussion_r1042501173) mentioned

dreambeyondorange commented on tests/integration-tests/tests/iam/test_iam.py:
```diff
@@ -10,16 +10,20 @@
 # This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied.
 # See the License for the specific language governing permissions and limitations under the License.
 import logging
+import os
```
looks like you are importing os twice?

himani2411 replied:
Will resolve this.

Commit:
Change according to PR

Update the pcluster.config.yaml file to remove Iam section and use _inject_resource_in_config() to inject Iam and ResourcePrefix Section
Change scope of initialize and register_prefix_cli_credentials from class to default(function) level
Update _test_iam_resource_in_cluster() to add Cluster creation verification
Update test_iam_resource_prefix to remove duplication of user-role-rp for each value of iam_resource_prefix_list test and improve performance.
Change position of user-role-iam-resource-prefix.cfn.yaml to the tests folder directory
Remove update_config variable from test_iam_resource_prefix

PR Link: https://github.com/aws/aws-parallelcluster/pull/4652

Commit:
Change Iam Resource Prefix tests to handle one test case

Add Iam/ResourcePrefix in pcluster.config.yaml
Change user-role-iam-resource-prefix.cfn.yaml to handle /path-prefix/name-prefix- Iam Resource Prefix
Add /path-prefix/name-prefix as the only pytest parameter and remove use_default_iam_credentials parameter check condition from initialize_resource_prefix_cli_creds

Commit:
Add Test specific CLI credentials in cluster creation

Add test specific CLI credentials while creation of cluster and run pcluster commands.
Change ClusterFactory's create_cluster() to have test specific CLI credentials as arguments
Change user-role-iam-resource-prefix.cfn.yaml to handle /path-prefix/name-prefix- Iam Resource Prefix
Remove register_resource_prefix_cli_credentials as it will affect parallel tests running in same region.

dreambeyondorange commented on tests/integration-tests/clusters_factory.py:
```diff
@@ -417,10 +424,12 @@ def create_cluster(self, cluster, log_error=True, raise_on_error=True, **kwargs)
 
         # create the cluster
         logging.info("Creating cluster {0} with config {1}".format(name, cluster.config_file))
+        self.custom_cli_credentials = kwargs.get("custom_cli_credentials")
```
Can you not just reference the credentials from the __init__ function?  Why reset this value here?

dreambeyondorange replied:
Either we should define the credentials for the factory or use the credentials provided to the create method, but not do both

dreambeyondorange replied:
I'm leaning toward the latter, since the factory can then be independent of the cluster's credentials and it seems that the factory only uses the credentials argument (and the run_pcluster_command) in the creation of the cluster

himani2411 replied:
Since the clusters_factory(scope class) fixture in conftest is calling the Clustersfactory and if I setup the custom_cli_credentials value there it will always take the default None value. As in test_iam class any class which comes first will setup the ClusterFactory object. So when I use cluster_factory fixture in my test case I would have to send my custom_credentials where it will reset the object member in my test case.

dreambeyondorange replied:
I'm not sure I follow.  It doesn't make sense to me to do both of these things
```
self.custom_cli_credentials = custom_cli_credentials
...
self.custom_cli_credentials = kwargs.get("custom_cli_credentials")
```
It seems like you are setting the class variable and getting the context from the create_cluster arguments.  So, the first time you set self.custom_cli_credentials is essentially dead code since it's replaced by the second call.  I think in this case since the clusters factory fixture is creating the Cluster object directly and bypassing the ClustersFactory class (a design flaw IMO) we should remove the credentials member variable from the ClustersFactory class and use the credentials provided to the create_cluster method.

himani2411 replied:
okay

dreambeyondorange commented on tests/integration-tests/framework/credential_providers.py:
```diff
@@ -25,15 +25,20 @@ def register_cli_credentials_for_region(region, iam_role):
 
-    if region in cli_credentials:
-        with sts_credential_provider(region, cli_credentials[region]):
+    if kwargs.get("custom_cli_credentials"):
+        with sts_credential_provider(region, credential_arn=kwargs.get("custom_cli_credentials")):
             return run_command(*args, **kwargs)
     else:
-        return run_command(*args, **kwargs)
+        if region in cli_credentials:
+            with sts_credential_provider(region, cli_credentials[region]):
+                return run_command(*args, **kwargs)
+        else:
+            return run_command(*args, **kwargs)
```
I think this switching would be better done on the credentials argument itself and avoid multiple return statements.  It should be possible to do this with one assignment
```
credentials_arg = kwargs.get("custom_cli_credentials") or cli_credentials.get(region)
```
the `or` should do the switching for you, i.e. if kwargs.get("custom_cli_credentials") returns None, `credentials_arg` will be the region credentials

I think we should also refactor the cli_credentials[region] to use the get() method to avoid missing key errors.

himani2411 replied:
Will make the changes.

dreambeyondorange commented on tests/integration-tests/tests/iam/test_iam.py:
```diff
@@ -332,3 +334,140 @@ def test_s3_read_write_resource(region, pcluster_config_reader, s3_bucket_factor
+        logging.warning("Skipping deletion of CFN stacks because --no-delete option is set")
+
+
+def _inject_resource_in_config(cluster_config, resource_value, resource_keys):
```
this looks like a candidate for a more general utility module.  Is this something that is specific to this test?

himani2411 replied:
I was planning on using this function for injecting my Resource for Iam_resource_prefix in the user-role.cfn.yaml which is discussed [here](https://github.com/aws/aws-parallelcluster/pull/4652#discussion_r1040909717). though I am planning on doing this in another PR.

himani2411 replied:
Will remove this as not being used in this PR but will include in another PR.

dreambeyondorange commented on tests/integration-tests/utils.py:
```diff
@@ -80,7 +80,16 @@ def retry_if_subprocess_error(exception):
+    timeout=None,
+    raise_on_error=True,
+    shell=False,
+    custom_cli_credentials=None,
```
It doesn't look like this argument is used?

himani2411 replied:
Run_command() definition didnt originally had variable arguments, so the kwargs that its being sent from run_pcluster_command now has custom_cli_credentials as one of the arguments. So either I mention the variable I want to send or just add **kwargs in the run_command definition to avoid an error. I preferred adding my own custom_cli_credentials in way of avoiding sending unnecessary variables through kwargs.

himani2411 replied:
Will change this after part of our discussion

Commit:
Remove unnecessary methods and keyword Arguments

Remove _inject_resource_in_config() which is unused for this PR( part of another PR)
Revert run_command() to its original definition
Update run_pcluster_command() to handle switching and KeyError for credential_arn
Remove unnecessary key word arguments passed from create_cluster() to run_pcluster_command()

Reference PR:
https://github.com/aws/aws-parallelcluster/pull/4652

Commit:
Remove custom_cli_credentials from ClusterFactory

Removing custom_cli_credentials from Cluster Factory class member.

Reference PR:
https://github.com/aws/aws-parallelcluster/pull/4652

dreambeyondorange commented on tests/integration-tests/framework/credential_providers.py:
```diff
@@ -25,15 +25,19 @@ def register_cli_credentials_for_region(region, iam_role):
 
-    if region in cli_credentials:
-        with sts_credential_provider(region, cli_credentials[region]):
+        if region in cli_credentials:
+            with sts_credential_provider(
+                region, credential_arn=kwargs.get("custom_cli_credentials") or cli_credentials.get(region)
+            ):
+                kwargs.pop("custom_cli_credentials", None)
+                return run_command(*args, **kwargs)
+        else:
```
did you mean for this to be indented under the `if not region`?

himani2411 replied:
Reverted the changes

Commit:
Revert change in run_pcluster_command

Revert the changes in indentation and scope of if clause in run_pcluster_command()

Reference PR:
https://github.com/aws/aws-parallelcluster/pull/4652
